cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+## User Impacting Changes
+
+- We no longer add `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_SUBSCRIPTION_ID` or `AZURE_PRINCIPAL_ID` to the `.env` file.
+- We no longer support environment variable substitution in the `.parameters.json` file in the `infra` folder, it is used just to configure static values.
+- In our templates we remove the `.parameters.json` files since they are no longer used (all values are user configured during first deployment). (*note: We can choose to keep these files if we think their existence provides value in our templates as a learning tool*)
+- We prompt and validate deployment parameters using the metadata applied in the bicep template.
+
+We can figure out how to rolls these changes out in a smooth enough way (for example, we could take the existence of a environment reference like `${FOO}` in a `.parameters.json` to mean we were on the "old" plan and `azd` can warn and do something sensible (or point folks at an article on how to upgrade).
+
+## Open Questions
+
+> How do we express that a deployment parameter should correspond to the environment id or environment name. Do we introduce types like `azd-env-id` and `azd-env-name` for our metadata on parameters? Should we instead have something like `azd-context` and that passes in a JSON object with well known keys like `id` and `name`?  Is there value in making it easy to flow the environment name (which now will often just be the string "default" into the infrastructure? Why would you use it vs `azd-env-id`?) 
+
+Let's experiment with a context object that will contain the environment id, as well as image ids for containers we've built and go from there 
+
+> Related to the above - how do we denote the "this parameter corresponds to the name of the of the image to deploy". In the older model we wrote a name with a well known key into the environment and then had the `.parameters.json` file wire it up to the deployment parameter. What do we want to do in this new world (where that doesn't work as well since we don't evaluate a `.parameters.json` file in an environment)? Something in an `azd-context` object?  Another `azd-image-name` type?
+
+We'll have a services object in the context which will have keys for each service (mirroring `azure.yaml`) and it will contain an `imageId` property.
+
+> How does this impact `azd env refresh` - What does refreshing an environment mean?  
+
+`env refresh` fetches the latest deployment object for the environment and updates `properties` and `outputs` sections of the `config.json` file for the `main` module. As a consequence of changing `config.json`, the `.env` and `.parameters.json` files are also regenerated.  (*note: I would honestly prefer if we just got rid of this command for now, I don't think we have a great understanding of what this tries to solve and the value it provides today is quite limited*)
```

Comment:
Do you mean remove `env refresh`?  This command is helpful when I want to share an environment with someone without having them provision it.  I can tell them to run `azd env refresh -e envname`

Reply:
It is also used by the tooling, big VS specifically, in order to determine what has been already provision, and whether there is a need to re-provision.

Reply:
Take a look at what `env refresh --output json` shows